### PR TITLE
Perftest/initial setup

### PR DESCRIPTION
**Pull request by cbeach47**
##### Checklist
-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [x] tests are included
-   [x] documentation is changed or added

##### Description of change
Setup an initial, general performance test for uPortal's ecosystem.  See the doc's README for details.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -57,3 +59,12 @@ apply plugin: 'java'
 compileJava {
     dependsOn javaVersionCheck
 }
+
+/*
+ * Enable JMeter testing
+ * TODO - enable:  jmSystemPropertiesFiles= [file("buildSrc/src/test/perf/smoke/smokeTest.properties")]
```
Comment (ChristianMurphy) [inline build.gradle]: @cbeach47 anything blocking enabling `jmSystemPropertiesFiles`?

File: docs/en/testing/perf/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+Test are run with a cache manager, but will clear the cache each iteration.
+
+1. Ramp up from 1 to `load.num.of.threads` users in `load.ramp.up` seconds
+2. All users perform the list of modules / actions below
+3. Between module requests, users will take a random length break (`delay.think.time.*`)
+4. At the end of the loop, users will take a random length break (`delay.final.think.time.*`)
+4. Test will run for each `load.loop.count` iterations or a duration of `scheduler.duration.seconds`
+
+### List of Baseline Modules
+* UI - Login
```
Comment (ChristianMurphy) [inline docs/en/testing/perf/README.md]: :art: 
```suggestion

* UI - Login
```

File: gradle/tasks/perf.gradle [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+task perfGenTestData() {
+    group 'Perf'
+    description 'Generates test data to prepare a uPortal install for performance testing'
+    dependsOn ':portalProperties'
+
+    doLast {
+        // Determine output location for user xml files
+        final String testDataDirKey = "data.test.perf.directory";
+        final String testDataDirStr = project.rootProject.ext['buildProperties'].getProperty(testDataDirKey)
+        if(testDataDirStr == null || testDataDirStr.length() == 0) {
```
Comment (ChristianMurphy) [inline gradle/tasks/perf.gradle]:
```suggestion
        if (testDataDirStr == null || testDataDirStr.length() == 0) {
```

File: gradle/tasks/perf.gradle [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            throw new GradleException("Please specify a [${testDataDirKey}] to generate the test data")
+        }
+
+        // Determine number of test users
+        final String testMaxUsersStr = project.rootProject.ext['buildProperties'].getProperty("data.test.perf.max.users");
+        int maxUsers = 1000 // Default
+        if(testMaxUsersStr != null && testMaxUsersStr.length() > 0) {
```
Comment (ChristianMurphy) [inline gradle/tasks/perf.gradle]:
```suggestion
        if (testMaxUsersStr != null && testMaxUsersStr.length() > 0) {
```

File: gradle/tasks/perf.gradle [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            maxUsers = Integer.valueOf(testMaxUsersStr)
+        }
+
+        // Determine location / name of population file
+        String testPopulationStr = project.rootProject.ext['buildProperties'].getProperty("data.test.perf.population.file");
+        if(testPopulationStr == null || testPopulationStr.length() == 0) {
```
Comment (ChristianMurphy) [inline gradle/tasks/perf.gradle]:
```suggestion
        if (testPopulationStr == null || testPopulationStr.length() == 0) {
```

File: gradle/tasks/perf.gradle [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            testPopulationStr = "buildSrc/src/test/perf/baseline-population.csv" // Default
+        }
+        File populationFile = file(testPopulationStr)
+
+        // Prep output location for user xml files
+        File testDataDir = file(testDataDirStr)
+        if(!testDataDir.exists()) {
```
Comment (ChristianMurphy) [inline gradle/tasks/perf.gradle]:
```suggestion
        if (!testDataDir.exists()) {
```

File: gradle/tasks/perf.gradle [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            testDataDir.mkdirs()
+            logger.lifecycle("Created perf test data directory ${testDataDir}")
+        } else {
+            logger.lifecycle("Using existing perf test data directory ${testDataDir}")
+        }
+
+        // Build user files and summarize in the population file (as an input to JMeter)
+        for(int i = 0; i < maxUsers; i++) {
```
Comment (ChristianMurphy) [inline gradle/tasks/perf.gradle]:
```suggestion
        for (int i = 0; i < maxUsers; i++) {
```

File: docs/en/testing/perf/README.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+As a proof-of-concept, you can execute the baseline scripts via Gradle.  The Gradle process is noted here as a possible CI hook for running perf tests at each commit / PR in the future.
+```sh
+$ ./gradlew jmRun
+```
+
+## Baseline Script
+
+Test are run with a cache manager, but will clear the cache each iteration.
+
```
[the remark this answers is not in the document]
Comment (cbeach47) [reply docs/en/testing/perf/README.md]: updated in new PR - https://github.com/Jasig/uPortal-start/pull/370

File: gradle/tasks/perf.gradle (hunk 7 above)
Comment (cbeach47) [reply gradle/tasks/perf.gradle]: updated in new PR - https://github.com/Jasig/uPortal-start/pull/370

File: gradle/tasks/perf.gradle (hunk 6 above)
Comment (cbeach47) [reply gradle/tasks/perf.gradle]: updated in new PR - https://github.com/Jasig/uPortal-start/pull/370

File: gradle/tasks/perf.gradle (hunk 5 above)
Comment (cbeach47) [reply gradle/tasks/perf.gradle]: updated in new PR - https://github.com/Jasig/uPortal-start/pull/370

File: gradle/tasks/perf.gradle (hunk 3 above)
Comment (cbeach47) [reply gradle/tasks/perf.gradle]: updated in new PR - https://github.com/Jasig/uPortal-start/pull/370

File: gradle/tasks/perf.gradle (hunk 4 above)
Comment (cbeach47) [reply gradle/tasks/perf.gradle]: updated in new PR - https://github.com/Jasig/uPortal-start/pull/370

File: docs/en/testing/perf/README.md (hunk 2 above)
Comment (cbeach47) [reply docs/en/testing/perf/README.md]: updated in new PR - https://github.com/Jasig/uPortal-start/pull/370

**Comment by cbeach47** [comment]
Lesson learned on 're forking' with open PRs.  New PR for this effort:  https://github.com/Jasig/uPortal-start/pull/370

**Pull request closed without merging**